[ITP] Refactor logic for managing the lifetimes of script-written third party cookies and local storage

whsieh opened the pull request:
#### 8defcbeb9bbafc2a135b9b5cd4703fdccb289335
<pre>
[ITP] Refactor logic for managing the lifetimes of script-written third party cookies and local storage
<a href="https://bugs.webkit.org/show_bug.cgi?id=280099">https://bugs.webkit.org/show_bug.cgi?id=280099</a>
<a href="https://rdar.apple.com/136378432">rdar://136378432</a>

Reviewed by Charlie Wolfe.

Add checks for script telemetry underneath calls to read from or write to `localStorage` or
`document.cookie`.

* Source/WebCore/loader/CookieJar.cpp:
(WebCore::CookieJar::getRawCookies const):
* Source/WebCore/loader/CookieJar.h:
* Source/WebCore/storage/Storage.cpp:
(WebCore::Storage::length const):
(WebCore::Storage::key const):
(WebCore::Storage::getItem const):
(WebCore::Storage::setItem):
(WebCore::Storage::removeItem):
(WebCore::Storage::requiresScriptExecutionTelemetry const):
* Source/WebCore/storage/Storage.h:
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::requiresScriptExecutionTelemetry):
(WebKit::WebCookieJar::cookies const):
(WebKit::WebCookieJar::setCookies):
(WebKit::WebCookieJar::cookiesEnabled):
(WebKit::WebCookieJar::cookieRequestHeaderFieldValue const):
(WebKit::WebCookieJar::getRawCookies const):
(WebKit::WebCookieJar::getCookiesAsync const):
(WebKit::WebCookieJar::setCookieAsync const):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView synchronouslyLoadSimulatedRequest:responseHTMLString:]):

Add a helper method to synchronously load a simulated `NSURLRequest`.

Canonical link: <a href="https://commits.webkit.org/284135@main">https://commits.webkit.org/284135@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f52e5d60675743abcf202446ff4d1be509e250a0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68538 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47930 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21197 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72607 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19683 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70655 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55726 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19499 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54673 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13091 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71605 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43808 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59181 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35136 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40475 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16604 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18040 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62429 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16951 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74301 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12509 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16207 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62142 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12549 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59260 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/62167 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15193 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10107 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3725 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43731 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44805 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45999 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44547 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
```diff
@@ -108,6 +108,7 @@ struct AutocorrectionContext {
 - (void)synchronouslyLoadHTMLString:(NSString *)html baseURL:(NSURL *)url;
 - (void)synchronouslyLoadHTMLString:(NSString *)html preferences:(WKWebpagePreferences *)preferences;
 - (void)synchronouslyLoadRequest:(NSURLRequest *)request;
+- (void)synchronouslyLoadSimulatedRequest:(NSURLRequest *)request responseHTMLString:(NSString *)htmlString;
```
I don’t think you meant to add this.

charliewolfe replied:
Oh never mind, you did (saw rdar).

charliewolfe commented on Source/WebCore/storage/Storage.cpp:
```diff
@@ -130,4 +146,10 @@ Ref<StorageArea> Storage::protectedArea() const
+bool Storage::requiresScriptExecutionTelemetry() const
+{
+    RefPtr document = window() ? window()->document() : nullptr;
+    return document->requiresScriptExecutionTelemetry(ScriptTelemetryCategory::LocalStorage);
```
null check?

whsieh replied:
Good catch!

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -270,6 +285,12 @@ std::pair<String, WebCore::SecureCookiesAccessed> WebCookieJar::cookieRequestHea
     RefPtr webFrame = frameID ? WebProcess::singleton().webFrame(*frameID) : nullptr;
+
+    if (RefPtr frame = webFrame ? webFrame->protectedCoreLocalFrame() : nullptr) {
+        if (RefPtr document = frame->protectedDocument(); document && requiresScriptExecutionTelemetry(*document))
```
```suggestion
        if (RefPtr document = frame->protectedDocument(); document && requiresScriptExecutionTelemetry(document.releaseNonNull()))
```

szewai reviewed (commented):
Is there a flag to turn off this behavior?

whsieh committed:
[ITP] Refactor logic for managing the lifetimes of script-written third party cookies and local storage
https://bugs.webkit.org/show_bug.cgi?id=280099
rdar://136378432

Reviewed by Charlie Wolfe.

Add checks for script telemetry underneath calls to read from or write to `localStorage` or
`document.cookie`.

* Source/WebCore/loader/CookieJar.cpp:
(WebCore::CookieJar::getRawCookies const):
* Source/WebCore/loader/CookieJar.h:
* Source/WebCore/storage/Storage.cpp:
(WebCore::Storage::length const):
(WebCore::Storage::key const):
(WebCore::Storage::getItem const):
(WebCore::Storage::setItem):
(WebCore::Storage::removeItem):
(WebCore::Storage::requiresScriptExecutionTelemetry const):
* Source/WebCore/storage/Storage.h:
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::requiresScriptExecutionTelemetry):
(WebKit::WebCookieJar::cookies const):
(WebKit::WebCookieJar::setCookies):
(WebKit::WebCookieJar::cookiesEnabled):
(WebKit::WebCookieJar::cookieRequestHeaderFieldValue const):
(WebKit::WebCookieJar::getRawCookies const):
(WebKit::WebCookieJar::getCookiesAsync const):
(WebKit::WebCookieJar::setCookieAsync const):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView synchronouslyLoadSimulatedRequest:responseHTMLString:]):

Add a helper method to synchronously load a simulated `NSURLRequest`.

Canonical link: https://commits.webkit.org/284135@main
